CLI: Component-driven React / Vue / Angular / Preact / Svelte templates

shilman opened the pull request:
Issue: #10723 

## What I did

- [x] Copy @dom & @tmeasday 's CLI template from https://github.com/chromaui/storybook-template-react
- [x] Update to match latest Args recommendations & minor tweaks
- [x] Rename `0.Introduction.stories.mdx` to `Introduction.stories.mdx`
- [x] Update `main.js` to order MDX first
- [x] Set up action logging out of box
- [x] Migrate `Button` to CSS

## What's next

- [ ] Discussion here (comments welcome!)
- [ ] Apply to other frameworks (help pls @tooppaaa @yannbf ?)

## How to test

Modify `scripts/run-e2e.ts`:
```
await exec(`node ../../storybook/lib/cli/dist/generate init`, { cwd });
// await exec(`npx -p @storybook/cli sb init --yes ${type}`, { cwd });
```

Then run:
```
yarn test:e2e-framework react
```

shilman committed:
CLI: Update React JS template with Header/Page stories

tmeasday commented on lib/cli/src/frameworks/react/js/Button.js:
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const baseStyles = {
```
I think we discussed putting these styles in an imported CSS file and using class names to switch between them. That way the amount of code we share between frameworks could be larger whilst still writing components in a "natural" way. 

WDYT @shilman?

tooppaaa replied:
I'm in favor of moving everything that is not tied to a framework in the common folder.

tmeasday commented on lib/cli/src/frameworks/react/js/Header.stories.js:
```diff
@@ -0,0 +1,18 @@
+
+import { Header } from './Header';
+
+export default {
+  title: 'Example/Header',
+  component: Header,
+};
```
Did we want to add `argTypes` action annotations or a `argTypeRegexp` parameter here?

tmeasday replied:
I guess in the template I ended up putting it in `preview.js` after some discussion. Do we want to do that here too? 

@yannbf mentioned that colleagues found it confusing when the docs mention putting stuff in `preview.js` (which you almost inevitably need to do in configuring SB), but the file isn't there to begin with.

OTOH there is something nice about having a minimal set of configuration files in place when you start and adding to them as you get more complex in your usage.

tmeasday replied:
https://github.com/chromaui/storybook-template-react/blob/c7eb09bf49da39b5e25226e7f6a69ba3a2b880af/.storybook/preview.js#L3

shilman replied:
Thanks for the quick review @tmeasday — will update!

tooppaaa replied:
> Do we want to do that here too?

It's already done when we introduced essentials in the CLI. 
In my opinion, I prefer having it in a single place and never think of it again :)

shilman committed:
Merge branch 'next' into 10723-update-cli-template-stories

shilman committed:
CLI: React template Button onClick

shilman committed:
CLI: React template Button css

shilman commented:
@tmeasday @tooppaaa @domyen I've updated the button to use CSS and it's working. If it looks good, I'll update the other ones. I'm not really up on CSS best practices.

@tmeasday the action stuff is already set up by @tooppaaa. i made it a little bit more specific and added an `onClick` propType to the template and it works.

tooppaaa committed:
add e2e target

tooppaaa commented:
E2E added that covers 
- should have Welcome story displayed
- should have 4 Button stories (check on classnames)
- should have 2 Header stories (check on button text content)
- should have 2 Page stories (check on button text + page content text)

I applied a bit of changes and extended your work @shilman on CSS.
Happy to make changes but the earliest possible as we can start working on others framework using E2E green as target !

domyen commented on lib/cli/src/frameworks/react/js/Page.js:
```diff
@@ -0,0 +1,75 @@
+      <p>
+        We recommend building UIs with a{' '}
+        <a
+          href="https://blog.hichroma.com/component-driven-development-ce1109d56c8e"
```
```suggestion
          href="http://componentdriven.org/"
```

@tmeasday can you set up componentdriven.org to redirect for now?

tmeasday replied:
Already did @domyen

domyen replied:
Is the redirect working for you? http://componentdriven.org

tmeasday replied:
That's weird it was on Friday

yannbf replied:
I tried all sorts of variations, only link that works is http://www.componentdriven.org/

tmeasday replied:
I think I fixed it.

shilman commented:
@tooppaaa These changes look great. 💯

yannbf commented on lib/cli/src/frameworks/react/ts/Button.tsx:
```diff
@@ -1,22 +1,48 @@
+      type="button"
+      className={['storybook-button', `storybook-button--${size}`, mode].join(' ')}
+      style={backgroundColor && { backgroundColor }}
+      {...props}
```
Won't typescript complain about this? ButtonProps does not contain all possible props for a button element, I might be wrong but I think you'd need to do something like:
```ts
export interface ButtonProps extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
    primary?: boolean;
    ....
}
```

yannbf commented on lib/cli/src/frameworks/react/ts/Header.stories.tsx:
```diff
@@ -0,0 +1,18 @@
+};
+
+export const LoggedOut = Template.bind({});
+LoggedOut.args = {};
```
Is the empty args here just to show people they can use it or is it necessary? There's already an example in `LoggedIn`, so no need to add it, right? It might lead to people thinking that they always need to add an empty args to every story

yannbf commented on lib/cli/src/generators/configure.ts:
```diff
@@ -20,7 +18,7 @@ function configureMain(addons: string[], custom?: any) {
   const parameters = `
 export const parameters = {
-  actions: { argTypesRegex: "^on.*" },
+  actions: { argTypesRegex: "^on[A-Z].*" },
```
do you think it would be nice to add a small comment telling people that they can change this to other stuff like `"^(on|handle)[A-Z].*"` in case they use a different pattern, or even already just add that regex because it might useful anyway?

yannbf commented on lib/cli/src/frameworks/react/ts/Page.tsx:
```diff
@@ -0,0 +1,71 @@
+      <ul>
+        <li>
+          Use a higher-level connected component. Storybook helps you compose such data from the
+          "args" of child component stories
```
maybe a hyperlink in the args would be useful for reference?

yannbf commented on lib/cli/src/frameworks/react/ts/Page.tsx:
```diff
@@ -0,0 +1,71 @@
+      <p>
+        We recommend building UIs with a{' '}
+        <a
+          href="https://blog.hichroma.com/component-driven-development-ce1109d56c8e"
```
new link for this: http://www.componentdriven.org/

yannbf commented:
I can confirm that http://componentdriven.org/ works now, not https though

-- 
Yann Braga
Software Developer
m: +31618338183
a: Amsterdam, Netherlands.
w: github.com/yannbf  e: yannbf@gmail.com
<https://linkedin.com/in/yannbraga>

tooppaaa committed:
Vue stories

shilman commented:
@tooppaaa @yannbf When I run the sample app (`react`, `react_typescript`) locally it looks like the CSS is not being applied properly. The "JS" props (`backgroundColor`, `label`, `onClick`) are working, but the CSS props (`primary`, `size`) are not. Tho the story is getting re-rendered and the DOM is changing AFAICT. I'm sure that this was working on my initial commits, but it doesn't seem to be working on rollback, so maybe it's due to a change elsewhere in SB .. or in CRA?

<img width="1030" alt="Example___Button_-_Small_⋅_Storybook" src="https://user-images.githubusercontent.com/488689/87365970-21c6a480-c5aa-11ea-8ea5-c0337da80f2f.png">

tmeasday commented:
We should probably get the https sorted too huh

tmeasday commented:
Fixed now, try https://componentdriven.org

shilman committed:
CLI template: Fix button CSS

shilman committed:
CLI template: Link to componentdriven.org

shilman committed:
Fix typo

shilman committed:
CLI: Ignore import/extensions for framework templates

tooppaaa committed:
Preact CLI

tooppaaa committed:
Angular CLI

shilman committed:
Merge branch 'next' into 10723-update-cli-template-stories

shilman committed:
CLI template: Fix componentdriven.org link

shilman committed:
CLI template: Use action argType instead of hard-coded action

shilman committed:
CLI template: Prettier

shilman committed:
CLI: Fix Angular template compodoc settings

shilman committed:
Merge branch 'next' into 10723-update-cli-template-stories

shilman committed:
Merge branch 'next' into 10723-update-cli-template-stories

shilman committed:
CLI: Fix Vue CLI template props handing

shilman commented:
I fixed a bunch of Vue/Angular bugs and this is pretty much ready to go.

Waiting on #11561 and discussion with @tmeasday to make sure this is the direction we want to go before merging, since it might be incompatible with storyshots.

tooppaaa committed:
Svelte CLI

shilman committed:
Merge branch 'next' into 10723-update-cli-template-stories

shilman committed:
Fix capitalization
